applications/FluidDynamicsApplication/python_scripts/navier_stokes_two_fluids_solver.py

```diff
@@ -0,0 +1,269 @@
+
+        rho_1 = properties_1.GetValue(KratosMultiphysics.DENSITY)
+        rho_2 = properties_2.GetValue(KratosMultiphysics.DENSITY)
+        mu_1 = properties_1.GetValue(KratosMultiphysics.DYNAMIC_VISCOSITY)
+        mu_2 = properties_2.GetValue(KratosMultiphysics.DYNAMIC_VISCOSITY)
+
+        if rho_1 <= 0.0:
+            raise Exception("DENSITY set to {0} in Properties {1}, positive number expected.".format(rho_1, properties_1.Id))
+        if rho_2 <= 0.0:
+            raise Exception("DENSITY set to {0} in Properties {1}, positive number expected.".format(rho_2, properties_2.Id))
+        if mu_1 <= 0.0:
+            raise Exception("DYNAMIC_VISCOSITY set to {0} in Properties {1}, positive number expected.".format(mu_1, properties_1.Id))
+        if mu_2 <= 0.0:
+            raise Exception("DYNAMIC_VISCOSITY set to {0} in Properties {1}, positive number expected.".format(mu_2, properties_2.Id))
+
+        # Transfer density and (dynamic) viscostity to the nodes
+        for node in self.main_model_part.Nodes:
+            if node.GetSolutionStepValue(KratosMultiphysics.DISTANCE) < 0.0:
+                node.SetSolutionStepValue(KratosMultiphysics.DENSITY, rho_1)
+                node.SetSolutionStepValue(KratosMultiphysics.DYNAMIC_VISCOSITY, mu_1)
+            else:
+                node.SetSolutionStepValue(KratosMultiphysics.DENSITY, rho_2)
+                node.SetSolutionStepValue(KratosMultiphysics.DYNAMIC_VISCOSITY, mu_2)
+
+    def _set_constitutive_law(self):
+        # Construct the two fluids constitutive law
+        if(self.main_model_part.ProcessInfo[KratosMultiphysics.DOMAIN_SIZE] == 3):
+            self.main_model_part.Properties[1][KratosMultiphysics.CONSTITUTIVE_LAW] = KratosCFD.NewtonianTwoFluid3DLaw()
+        elif(self.main_model_part.ProcessInfo[KratosMultiphysics.DOMAIN_SIZE] == 2):
+            self.main_model_part.Properties[1][KratosMultiphysics.CONSTITUTIVE_LAW] = KratosCFD.NewtonianTwoFluid2DLaw()
+
+    def _set_distance_function(self):
+        ## Set the nodal distance function
+        if (self.settings["distance_reading_settings"]["import_mode"].GetString() == "from_GiD_file"):
+            import read_distance_from_file
+            DistanceUtility = read_distance_from_file.DistanceImportUtility(self.main_model_part, self.settings["distance_reading_settings"])
+            DistanceUtility.ImportDistance()
+        elif (self.settings["distance_reading_settings"]["import_mode"].GetString() == "from_mdpa"):
+            KratosMultiphysics.Logger.PrintInfo("Navier Stokes Embedded Solver","Distance function taken from the .mdpa input file.")
+            # Recall to swap the distance sign (GiD considers d<0 in the fluid region)
+            for node in self.main_model_part.Nodes:
+                distance_value = node.GetSolutionStepValue(KratosMultiphysics.DISTANCE)
+                node.SetSolutionStepValue(KratosMultiphysics.DISTANCE, -distance_value)
+
```

Comment:
Why is this necessary? For both GID and TwoFluids formulation, the "liquid" is in the negative side, (the other is usually air).

Reply:
This comes from the embedded solver. If we use the distance function generated by the GiD octree mesher, the fluid domain has negative distance sign while the structure domain has positive one. The one-fluid embedded formulations have the opposite sign criterion, so we swap the level set sign in this particular case.

However, as you say this is not the case for the two-fluids solver. So it can be removed. In any case, we have never used the GiD distance computation in combation with the two fluids solver (as far as I know).

Reply:
Indeed, everything is working because we set the distance in the main script, so no memory leaks appear. Once we somehow set the distance from a GUI, definitively we will need to rearrange this part of the solver. 

Thus, what we really need to think about is in the way we initialize the distance.

Reply:
Yes, this part has not become relevant so far due to the initialization of the distance field in MainKratos.py. The discussed lines were removed.